src/Build/Evaluation/LazyItemEvaluator.cs

```diff
@@ -342,15 +338,64 @@ private static ImmutableList<ItemData>.Builder ComputeItems(LazyItemList lazyIte
 
                 ImmutableHashSet<string> currentGlobsToIgnore = globsToIgnoreStack == null ? globsToIgnore : globsToIgnoreStack.Peek();
 
+                Dictionary<string, UpdateOperation> itemsWithNoWildcards = new Dictionary<string, UpdateOperation>(FileUtilities.GetIsFileSystemCaseSensitive() ? StringComparer.Ordinal : StringComparer.OrdinalIgnoreCase);
+                bool addedToBatch = false;
+
                 //  Walk back down the stack of item lists applying operations
                 while (itemListStack.Count > 0)
                 {
                     var currentList = itemListStack.Pop();
 
+                    if (currentList._memoizedOperation.Operation is UpdateOperation op)
+                    {
+                        bool addToBatch = true;
+                        int i;
+                        for (i = 0; i < op.ISpec.Fragments.Count; i++)
+                        {
+                            ItemSpecFragment frag = op.ISpec.Fragments[i];
+                            if (MSBuildConstants.CharactersForExpansion.Any(frag.TextFragment.Contains))
+                            {
+                                // Fragment contains wild cards, items, or properties. Cannot batch over it using a dictionary.
+                                addToBatch = false;
+                                break;
+                            }
+
+                            string fullPath = FileUtilities.GetFullPath(frag.TextFragment, frag.ProjectDirectory);
+                            if (itemsWithNoWildcards.ContainsKey(fullPath))
+                            {
+                                // Another update will already happen on this path. Make that happen before evaluating this one.
+                                addToBatch = false;
```

Comment:
Is this the right approach, or should it be a dictionary of lists of operations? This basically falls back to the current behavior, right?

Reply:
I think a dictionary of lists of operations would be asymptotically better, but with exactly 0 evidence, it feels unusual to have a large number of item updates to the same fragment in a row with nothing between them. Note that each time this happens, it resets, so we evaluate the current batch and move on; it doesn't prevent another mini-batch later. Having lists of operations would also be more complicated and noticeably increase allocations (a list for every fragment without a wildcard in an update operation), so I don't think it's worth it. I could do something in-between and immediately evaluate the mini-batch when I see a duplicate and add it to the next mini-batch, but I'm not sure it's worth it.